proposals/filecoin-retrieval-in-ipfs.md

```diff
@@ -0,0 +1,163 @@
+
+#### Leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+🎯🎯🎯
+
+While the initial project would only aim to support only the free-retrieval use case, there's a clear path to all kinds of paid retrievals. It would help us flesh out key issues around discovery for retrievals. (for example extending what goes in the IPFS DHT, learning how to decide between free or paid retrieval). It should at minimum make things easier for pinning services who want to use Filecoin.
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+High
+
+We know multiple companies that wish to use Filecoin as the storage provider for IPFS data. We know that many have been unsuccessful so far in bridging the gap.
+
+## Project definition
+#### Brief plan of attack
+
+1. We work with libp2p & IPFS stewards to come up with design for how we put filecoin retrieval records in the DHT. Note: I have used language so far that assumes these records will go in the IPFS DHT itself, but it may make sense to explore putting them in a seperate DHT that IPFS also has access too.
+
+2. We work out a way in Lotus to get these DHT records posted as part of the deal making process.
+
+3. We add the data transfer module to IPFS and the retrieval portion of go-fil-markets backed by a node interface that errors on chain based requests.
+
+4. We implement a negotiation mechanism for deciding between Filecoin retrieval via data transfer vs traditional bitswap in IPFS (note: this requires significant design). This negotiation happens in the [currently-in-development rewritten go-ipfs fetcher](https://github.com/ipfs/go-fetcher). Alternatively, perhaps data transfer itself becomes the top level interface above the fetcher? 
+
+5. We work with select pinning providers (eg Pinata) to verify we've provided the components for roundtrip pinning to Filecoin from IPFS
+
+6. We design the data transfer plugin API in concert with those who are looking at providing alternate mechanisms of retrieval (note: it's not clear to me if this should be a local plugin system or an HTTP API -- a local plugin can do any kind of custom HTTP communication, but is also harder to install and configure)
+
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+Done is when you can complete a roundtrip pinning / retrieval to Filecoin from IPFS and a spec exists for building future paid retrievals
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+- Well known pinning services begin offering Filecoin based pinning
```

Comment:
This isn't strictly required as you can pin directly to a Filecoin light client, although it's a nice way to see if things are catching on.

Reply:
Where is the spec for what Filecoin light clients can do? My hope with light clients is that they are purely lightweight interfaces to the Filecoin network and don't do any direct storage/pinning themselves

Reply:
There isn't one in particular that I know of. #40 is sort of there but is likely heavier than what you'd expect.

I wasn't being very clear here, so I'll try and clarify. My thought was that some reasonably light Filecoin client could be asked to store data on Filecoin using the remote pinning API, it would download the data from IPFS and then send it off to the miner who is ultimately storing the data. At that point the Filecoin client could just delete the data. If the Filecoin client + IPFS nodes are on the same machine this seems like it could be pretty efficient.